src/fat_beagle.cpp

```diff
@@ -511,16 +530,41 @@ std::vector<double> RatioGradient(const RootedTree &tree,
   return gradientLogDensity;
 }
 
+std::vector<double> DiscreteSiteModelGradient(
+    const std::vector<double> &branch_lengths,
+    const std::vector<double> &unscaled_category_gradient) {
+  size_t edge_count = branch_lengths.size() - 1;
+  double rate_gradient = 0;
+  for (size_t node_id = 0; node_id < edge_count; node_id++) {
+    rate_gradient += unscaled_category_gradient[node_id] * branch_lengths[node_id];
+  }
+  return {rate_gradient};
+}
+
 UnrootedTreeGradient FatBeagle::Gradient(const UnrootedTree &in_tree) const {
   auto tree = in_tree.Detrifurcate();
   tree.SlideRootPosition();
+  EigenMatrixXd dQ = GetBranchDifferentialMatrices(
+      *phylo_model_->GetSiteModel(), *phylo_model_->GetSubstitutionModel());
   auto [log_likelihood, branch_length_gradient] =
-      BranchGradientInternals(tree.Topology(), tree.BranchLengths());
-  // We want the fixed node to have a zero gradient.
-  branch_length_gradient[tree.Topology()->Children()[1]->Id()] = 0.;
+      BranchGradientInternals(tree.Topology(), tree.BranchLengths(), dQ);
 
   std::vector<double> substitution_model_gradient;
   std::vector<double> site_model_gradient;
+  auto site_model = phylo_model_->GetSiteModel();
```

Comment:
This block and the below can get factored out to reduce code duplication as well, right?